File: src/web/rows.ts
/**
 * Objects → table rows.
 *
 * Extracted so `/api/list` and the live watch stream cannot drift: a row built
 * from a watch event must be byte-identical to the same row built from a list,
 * or a table that switches between the two would flicker in ways nobody could
 * reproduce.
 */

import type { K8sObject } from "../kubectl.ts";
import type { KindDef } from "../format.ts";
import { ingressRoutes } from "../format.ts";
import { type Meter, type UsageColumns } from "./gauges.ts";

export interface Row {
    cells: string[];
    name: string;
    ns?: string;
    tones: Record<number, string>;
    ts: number;
    labels?: Record<string, string>;
    meters?: Record<number, Meter>;
    rules?: { url: string; host: string; path: string; service: string; port: string }[];
    lines?: number;
}

export function statusTone(value: string): "ok" | "warn" | "bad" | "neutral" {
    const v = value.toLowerCase();
    if (v === "running" || v === "active" || v === "ready" || v === "bound" || v === "succeeded" || v === "true") {
        return "ok";
    }
    if (v === "pending" || v === "containercreating" || v === "terminating" || v === "notready" || v === "false") {
        return "warn";
    }
    if (v.includes("error") || v.includes("failed") || v === "crashloopbackoff" || v === "evicted" || v === "oomkilled") {
        return "bad";
    }
    return "neutral";
}

/**
 * Where the usage columns get spliced in, and what the header ends up as.
 *
 * Usage sits inside the kind's own columns rather than after them, so CPU
 * reads next to the replica counts and before AGE, where an operator looks.
 */
export function columnsFor(kind: KindDef, usage: UsageColumns | null): { columns: string[]; insertAt: number } {
    const base = kind.clusterScoped ? [...kind.columns] : ["NAMESPACE", ...kind.columns];
    if (!usage) return { columns: base, insertAt: base.length };
    const found = base.indexOf(usage.insertBefore);
    const insertAt = found >= 0 ? found : base.length;
    return { columns: [...base.slice(0, insertAt), ...usage.columns, ...base.slice(insertAt)], insertAt };
}

export function buildRow(
    obj: K8sObject,
    kind: KindDef,
    usage: UsageColumns | null,
    insertAt: number,
): Row {
    const name = obj.metadata?.name ?? "";
    const ns = obj.metadata?.namespace;
    const base = kind.row(obj);
    const withNs = kind.clusterScoped ? base : [ns ?? "", ...base];
    const extra = usage?.byKey.get(`${ns ?? ""}/${name}`);
    const cells = extra
        ? [...withNs.slice(0, insertAt), ...extra.cells, ...withNs.slice(insertAt)]
        : usage
          ? // A row the metrics pass never saw (it appeared between the list and
            // the top call) keeps its place in the grid with em dashes rather
            // than shifting every column left.
            [...withNs.slice(0, insertAt), ...usage.columns.map(() => "—"), ...withNs.slice(insertAt)]
          : withNs;

    const meters: Record<number, Meter> = {};
    extra?.meters.forEach((m, i) => {
        if (m) meters[insertAt + i] = m;
    });

    const tones: Record<number, string> = {};
    cells.forEach((c, i) => {
        const t = statusTone(c);
        if (t !== "neutral") tones[i] = t;
    });

    // An Ingress row's whole meaning is its routes, so they travel as
    // structured links: the URL is openable and the backend clickable, instead
    // of a comma-joined string you have to retype.
    const rules = kind.name === "ingresses" ? ingressRoutes(obj).slice(0, 12) : undefined;

    return {
        cells,
        name,
        ns,
        tones,
        // The creation timestamp travels with the row so the grid can sort by
        // real age. Sorting the rendered "5d" string puts 9m after 5d, which is
        // the kind of wrong that looks right.
        ts: obj.metadata?.creationTimestamp ? new Date(obj.metadata.creationTimestamp).getTime() : 0,
        labels: obj.metadata?.labels,
        meters: Object.keys(meters).length ? meters : undefined,
        rules,
        // A multi-route Ingress needs a taller row; the grid lays rows out by
        // their own heights.
        lines: rules && rules.length > 1 ? Math.min(rules.length, 4) : undefined,
    };
}

export const rowKey = (row: { ns?: string; name: string }): string => `${row.ns ?? ""}/${row.name}`;

/** Cheap identity for "did this row actually change?" delta suppression. */
export const rowFingerprint = (row: Row): string => JSON.stringify(row);

File: go/internal/server/rows.go
package server

import (
	"encoding/json"
	"strings"

	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"

	"github.com/notshekhar/digg/internal/model"
)

// Objects → table rows. Port of src/web/rows.ts.
//
// Extracted so /api/list and the live watch stream cannot drift: a row built
// from a watch event must be byte-identical to the same row built from a list,
// or a table that switches between the two would flicker in ways nobody could
// reproduce.

// Row is one line of a resource table.
type Row struct {
	Cells []string       `json:"cells"`
	Name  string         `json:"name"`
	NS    string         `json:"ns,omitempty"`
	Tones map[int]string `json:"tones"`
	// TS is the creation timestamp in ms, so the grid can sort by real age.
	// Sorting the rendered "5d" string puts 9m after 5d, which is the kind of
	// wrong that looks right.
	TS     int64                `json:"ts"`
	Labels map[string]string    `json:"labels,omitempty"`
	Meters map[int]*Meter       `json:"meters,omitempty"`
	Rules  []model.IngressRoute `json:"rules,omitempty"`
	// Lines makes a multi-route Ingress taller; the grid lays rows out by their
	// own heights.
	Lines int `json:"lines,omitempty"`
}

// StatusTone colours a cell by what its text means.
func StatusTone(value string) string {
	v := strings.ToLower(value)
	switch v {
	case "running", "active", "ready", "bound", "succeeded", "true":
		return "ok"
	case "pending", "containercreating", "terminating", "notready", "false":
		return "warn"
	case "crashloopbackoff", "evicted", "oomkilled":
		return "bad"
	}
	if strings.Contains(v, "error") || strings.Contains(v, "failed") {
		return "bad"
	}
	return "neutral"
}

// ColumnsFor says where the usage columns get spliced in, and what the header
// ends up as.
//
// Usage sits inside the kind's own columns rather than after them, so CPU reads
// next to the replica counts and before AGE, where an operator looks.
func ColumnsFor(kind *model.KindDef, usage *UsageColumns) ([]string, int) {
	base := make([]string, 0, len(kind.Columns)+1)
	if !kind.ClusterScoped {
		base = append(base, "NAMESPACE")
	}
	base = append(base, kind.Columns...)

	if usage == nil {
		return base, len(base)
	}
	insertAt := len(base)
	for i, c := range base {
		if c == usage.InsertBefore {
			insertAt = i
			break
		}
	}
	out := make([]string, 0, len(base)+len(usage.Columns))
	out = append(out, base[:insertAt]...)
	out = append(out, usage.Columns...)
	out = append(out, base[insertAt:]...)
	return out, insertAt
}

const maxIngressRules = 12

// BuildRow renders one object as a row.
func BuildRow(o *model.Obj, kind *model.KindDef, usage *UsageColumns, insertAt int) Row {
	name, ns := o.GetName(), o.GetNamespace()

	withNS := kind.Row(o)
	if !kind.ClusterScoped {
		withNS = append([]string{ns}, withNS...)
	}

	cells := withNS
	meters := map[int]*Meter{}
	if usage != nil {
		extra, ok := usage.ByKey[ns+"/"+name]
		if !ok {
			// A row the metrics pass never saw (it appeared between the list and
			// the top call) keeps its place in the grid with em dashes rather
			// than shifting every column left.
			blanks := make([]string, len(usage.Columns))
			for i := range blanks {
				blanks[i] = "—"
			}
			extra = usageCell{Cells: blanks}
		}
		out := make([]string, 0, len(withNS)+len(extra.Cells))
		out = append(out, withNS[:insertAt]...)
		out = append(out, extra.Cells...)
		out = append(out, withNS[insertAt:]...)
		cells = out

		for i, m := range extra.Meters {
			if m != nil {
				meters[insertAt+i] = m
			}
		}
	}

	tones := map[int]string{}
	for i, c := range cells {
		if t := StatusTone(c); t != "neutral" {
			tones[i] = t
		}
	}

	row := Row{Cells: cells, Name: name, NS: ns, Tones: tones, Labels: o.GetLabels()}

	if ts := o.GetCreationTimestamp(); !ts.IsZero() {
		row.TS = ts.UnixMilli()
	}
	if len(meters) > 0 {
		row.Meters = meters
	}

	// An Ingress row's whole meaning is its routes, so they travel as
	// structured links: the URL is openable and the backend clickable, instead
	// of a comma-joined string you have to retype.
	if kind.Name == "ingresses" {
		rules := model.IngressRoutes(o)
		if len(rules) > maxIngressRules {
			rules = rules[:maxIngressRules]
		}
		row.Rules = rules
		if len(rules) > 1 {
			row.Lines = min(len(rules), 4)
		}
	}
	return row
}

// RowKey identifies a row across list and watch.
func RowKey(ns, name string) string { return ns + "/" + name }

// RowFingerprint is a cheap identity for "did this row actually change?" delta
// suppression.
func RowFingerprint(r Row) string {
	b, err := json.Marshal(r)
	if err != nil {
		return ""
	}
	return string(b)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// ── small unstructured helpers shared by this package ───────────────────────

func nestedString(o *model.Obj, path ...string) string {
	v, ok, err := unstructured.NestedString(o.Object, path...)
	if err != nil || !ok {
		return ""
	}
	return v
}

func nestedMap(o *model.Obj, path ...string) map[string]any {
	v, ok, err := unstructured.NestedMap(o.Object, path...)
	if err != nil || !ok {
		return nil
	}
	return v
}

func nestedStringMap(o *model.Obj, path ...string) map[string]string {
	v, ok, err := unstructured.NestedStringMap(o.Object, path...)
	if err != nil || !ok {
		return nil
	}
	return v
}

func nestedSlice(o *model.Obj, path ...string) []any {
	v, ok, err := unstructured.NestedSlice(o.Object, path...)
	if err != nil || !ok {
		return nil
	}
	return v
}

func nestedInt(o *model.Obj, path ...string) int64 {
	v, ok, err := unstructured.NestedFieldNoCopy(o.Object, path...)
	if err != nil || !ok {
		return 0
	}
	switch n := v.(type) {
	case int64:
		return n
	case float64:
		return int64(n)
	}
	return 0
}

func nestedBool(o *model.Obj, path ...string) bool {
	v, ok, err := unstructured.NestedBool(o.Object, path...)
	if err != nil || !ok {
		return false
	}
	return v
}
